### ISSUE TITLE
full optimization can have nasty side effects when calling functions

### ISSUE THREAD

**Issue by deltanedas** [rhaiscript/rhai#749]
take this function for example:
```
engine.register_fn("info", |s: &str| log::info!("{s}"));
```

with `OptimizationLevel` set to `Full` when compiling a script using `engine.compile_file(path)` it will evaluate a function, even if it's unreachable
in this case the side effect is printing things while compiling the AST, but it could be worse.

say if i had a rhai script controlling a nuclear missile and i provided it a `launch()` function, then my `if code_correct && president_okayed && whatever { launch(); }` is completely ignored and it launches anyway *just by compiling it*, not very good.

in my real usecase (not a missile thank god) i am currently logging stuff inside a callback so its jarring to see messages while compiling, not even touching logic yet. a caveat of the bug is that it will only try to call when the arguments are constant, so `info(`hi ${param}`);` will not be called, but `info("hello");` will.

basically there should be a way to mark a function as impure or something that the optimizer never tries to call it.

great library btw

**Comment by schungx** [comment]
Yes, full optimization is dangerous, as is fully documented in the book.

However, functions are only run with constant arguments. Non-constant arguments do not trigger function calls...

Although functions with no arguments may be a prob...

I would suggest you stay away from full optimization. It is intended for engines that do not yet have functions added to them, as documented in the book.

It is also less useful than before because lots of common operators are now eagerly evaluated even for simple optimizations. So that takes away the original purpose of full optimization.

**Comment by schungx** [comment]
Use full optimization only when you have functions that are pure and you use those for clearer code.

For example:

```rust
let id = make_id(123, "hello", true);
```

If `make_id` is deterministic but is complicated and you don't want to just write out the result, which may be a complex type.  Then you can use full optimization to good effect.
